program/src/processor.rs

```diff
@@ -2311,6 +2313,385 @@ impl Processor {
         Ok(())
     }
 
+    #[inline(never)]
+    fn edit_spot_order(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        cancel_order: serum_dex::instruction::CancelOrderInstructionV2,
+        expected_cancel_size: u64,
+        mut new_order: serum_dex::instruction::NewOrderInstructionV3,
+    ) -> MangoResult<()> {
+        const NUM_FIXED: usize = 22;
+        let (fixed_ais, packed_open_orders_ais) = array_refs![accounts, NUM_FIXED; ..;];
+        let [
+            mango_group_ai,         // read
+            mango_account_ai,       // write
+            owner_ai,               // read & signer
+            mango_cache_ai,         // read
+            dex_prog_ai,            // read
+            spot_market_ai,         // write
+            bids_ai,                // write
+            asks_ai,                // write
+            _dex_request_queue_ai,   // write
+            dex_event_queue_ai,     // write
+            dex_base_ai,            // write
+            dex_quote_ai,           // write
+            base_root_bank_ai,      // read
+            base_node_bank_ai,      // write
+            base_vault_ai,          // write
+            quote_root_bank_ai,     // read
+            quote_node_bank_ai,     // write
+            quote_vault_ai,         // write
+            token_prog_ai,          // read
+            signer_ai,              // read
+            dex_signer_ai,          // read
+            _msrm_or_srm_vault_ai,   // read
+        ] = fixed_ais;
+
+        let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        check_eq!(token_prog_ai.key, &spl_token::ID, MangoErrorCode::InvalidProgramId)?;
+        check_eq!(dex_prog_ai.key, &mango_group.dex_program_id, MangoErrorCode::InvalidProgramId)?;
+        check!(signer_ai.key == &mango_group.signer_key, MangoErrorCode::InvalidSignerKey)?;
+
+        let market_index = mango_group
+            .find_spot_market_index(spot_market_ai.key)
+            .ok_or(throw_err!(MangoErrorCode::InvalidMarket))?;
+        let market_open_orders_ai = &packed_open_orders_ais[market_index];
+
+        // fetch the size of the order that the user is cancelling
+        let mut remaining_cancel_size: u64 = 0;
+        {
+            let open_orders = load_open_orders(market_open_orders_ai)?;
+
+            let market = load_market_state(spot_market_ai, dex_prog_ai.key)?;
+            let bids = load_bids_mut(&market, bids_ai)?;
+            let asks = load_asks_mut(&market, asks_ai)?;
+
+            for j in 0..128 {
+                let slot_mask = 1u128 << j;
+                if open_orders.free_slot_bits & slot_mask != 0 {
+                    // means slot is free
+                    continue;
+                }
+                let order_id = open_orders.orders[j];
+
+                if open_orders.is_bid_bits & slot_mask != 0 {
```

Comment:
why even call `find_by_key()` for orders that don't have the same side as `new_order.side` or `cancel_order.side`?